### ISSUE TITLE
CMake Build Pull Requests - Discussion

### ISSUE THREAD

**Issue by rodneyp290** [open-dis/open-dis-cpp#36]
I saw the recent CMake Pull Request #35, and after seeing @leif81's comments and questions, as well as the previously unaccepted CMake pull request #33, I thought it would be useful to create an issue for it to be discussed. 
One query from Lief was if CMake worked well with Visual Studio, and in the past I have found it was quite well across platforms.
The second query was if CMake is typically verbose?
This is were I'm a little unsure on an answer because I'm kind of a CMake newbie myself, and I originally though it wasn't so verbose.
My impression is that it depends what you want it to do, and I'm hoping that is what this issue can clarify.
Here is a simplistic 50 line [CMakeLists.txt](https://github.com/open-dis/open-dis-cpp/files/4095317/CMakeLists.txt) that I wrote up a while back.
It creates 2 separate libraries OpenDIS7 & OpenDIS6, and 2 executable files using ExampleSender & ExampleReciever.
It also includes an install target that puts the libraries in an `${CMAKE_INSTALL_PREFIX}/lib`, executables in `${CMAKE_INSTALL_PREFIX}/bin`, and copies the head files to ``${CMAKE_INSTALL_PREFIX}/include`.
the `${CMAKE_INSTALL_PREFIX}` defaults to `/usr/local/` on my linux system, however it can be redirected by setting it in the cmake command. So building and installing might look something like this (Assuming the attached CMakeList.txt is in the root directory of open-dis-cpp)
```sh
cd open-dis-cpp
mkdir build 
cd build
cmake -DCMAKE_INSTALL_PREFIX=/alternate/install/path/
make install # adding sudo if needed
```

**However, some issues with my simplistic CMakeList.txt include:**
- it doesn't check that you have dependencies such as SDL/SDL_net for the examples like #35 
- it doesn't use CPACK to a deb (#35) , or rpm (#33) package file
- it doesn't separate the utils into a separate library like both #33 & #35
- it doesn't separate the utils into a separate library like both #33 & #35
- it doesn't have/create the necessary *Config.cmake files that enable easy inclusion in other CMake projects like both #33 & #35 

So to sum up I think they needs to be a discussion, and decision on what requirements of the build system are, before a CMakeList.txt file is merged.

**Comment by rodneyp290** [comment]
Some key decisions:
How many libraries are required? 
  - [ ] OpenDIS6
  - [ ] OpenDIS7
  - [ ] OpenDIS-utils
What distro package files should be created?
  - [ ] Debian derivatives (.deb)
  - [ ] RedHat derivatives (.rpm)
Following the last question, what should be packaged?
  - Libraries, yes
  - header files, yes
  - Example Applications, probably not. **NOTE:** This is the only thing with the SDL dependencies
Are the *Config.cmake files that enable easy inclusion in other CMake projects required?
  - Probably a good idea

And lastly just a curiosity query from myself, if more than one CMake files is needed, how difficult would it be to separate it from the src tree? The thinking being that it might be more manageable to maintain.

**Comment by leif81** [comment]
Hi @rodneyp290 thank-you for doing this analysis.

Your simplistic 50 line CMakeLists.txt has me interested. I would feel good about using this as the starting point for migrating from Premake to CMake and then addressing the other limitations you identified as separate issues and PR's where we can have a discussion.

Would you mind turning your simple CMake file into a PR and go so far as getting the Travis build going with it and updating the README.md to demonstrate how to build with it on Linux and Windows?

**Comment by rodneyp290** [comment]
Okay, working on it at https://github.com/rodneyp290/open-dis-cpp/tree/cmake.
I think I just gotta update the README - I was able to add in the CPack package target, and it is still only 63 lines.
I might also add some comments to the CMakeList.txt to make it easier to understand.

**Comment by leif81** [comment]
Sounds good! When it's ready open a pull request here and I'll take another look.

**Comment by rodneyp290** [comment]
Hey, apologies for delays. Just a quick update.

I tried to write the instructions a while ago, and thought I better actually run it on window / visual studio. Thinks didn't quite work as nicely as expected.

Since then I've only recently, got around to debugging and fixing up the windows issues - and then fixes issues caused in linux, and ensure making sure it builds on travis-ci , and repeating a few times.
And finally, I believe it is working on all three.

Now... I just have to write those instructions... And hopefully I'll have a pull request up soon 😅

**Comment by leif81** [comment]
Ok thank you for the update @rodneyp290 . Your effort is appreciated, a number of people in the community are interested in CMake now.

**Comment by leif81** [comment]
Closed by #37 which was just merged in.
